### PR TITLE
Date Range after iD update

### PR DESCRIPTION
**Pull request by gregallensworth**
Refer to https://github.com/OpenHistoricalMap/issues/issues/395

- reinstates the Date Range panel into the new iD
- enhances Date Range to accept yyyy-mm-dd dates, and brings in decimaldate to compare day-level dates

### CODE REVIEW

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js]: #137 added a `utilDatesOverlap()` method to util.js that would allow this to say something like:

```js
const withinrange = utilDatesOverlap(selectedRange, entityRange);
```

without having to convert to decimal dates first. `utilDatesOverlap()` assumes dates are well-formed so that it can simply perform some string comparisons. I assume these string comparisons would be more performant than converting to decimal dates first. If a date is malformed, it treats it like the empty string and assumes the feature stretches to the beginning or end of time.

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js]: The new validator in #138 handles overflow and padding too. It adds another util.js method that could come in handy here:

```js
const selectedRange = {
  'start_date': utilNormalizeDateString(context.features().dateRange[0]).value,
  'end_date': utilNormalizeDateString(context.features().dateRange[1]).value
};
```

File: modules/ui/sections/map_daterange.js
Comment (1ec5) [inline modules/ui/sections/map_daterange.js:6]: I’d be in favor of relaxing this constraint at some point, so that users don’t get trapped in a situation where the feature they drew is no longer accessible in iD and it isn’t clear why: OpenHistoricalMap/issues#459. But this constraint was in the previous implementation, so I can bring this up in a separate issue.

File: modules/ui/sections/map_daterange.js
Comment (1ec5) [inline modules/ui/sections/map_daterange.js]: In the future, maybe we could consider replacing this separate section with a selectively disclosed row right under the “OpenHistoricalMap Data” checkbox in the Data Layers section. This would make the date filter more discoverable and also consistent with the [street-level imagery date filter](https://github.com/openstreetmap/iD/blob/e9bdd79887c5c8272c3ac9d345aa10943adaf275/modules/ui/sections/photo_overlays.js#L214-L231).

File: modules/ui/sections/map_daterange.js
Comment (1ec5) [inline modules/ui/sections/map_daterange.js]: openstreetmap/iD#8817 replaced `t()` with `t.append()` to mitigate against cross-site scripting attacks. Now a call to `t()` will result in an error.

File: modules/ui/sections/map_daterange.js
Comment (gregallensworth) [reply modules/ui/sections/map_daterange.js:6]: This min/max is not enforced by the Date Range boxes, is just the starting/default values. I should rename them `DEFAULT_START_DATE` and `DEFAULT_END_DATE`, since they're what's in the box and not an enforced min/max.

Comment (1ec5) [reply modules/ui/sections/map_daterange.js:6]: Great, then that pretty much addresses the issue I reported in OpenHistoricalMap/issues#459.

File: modules/renderer/features.js
Comment (gregallensworth) [reply modules/renderer/features.js]: I have changed to use `utilNormalizeDateString()` and `utilDatesOverlap()`

File: modules/renderer/features.js
Comment (gregallensworth) [reply modules/renderer/features.js]: I have changed to use `utilNormalizeDateString()` and `utilDatesOverlap()`

File: modules/ui/sections/map_daterange.js
Comment (gregallensworth) [reply modules/ui/sections/map_daterange.js]: Changed.

File: data/core.yaml
Comment (1ec5) [inline data/core.yaml]:
```suggestion
      tooltip: The start date in YYYY-MM-DD format. Negative years OK.
```

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js]: I forgot to mention that `utilNormalizeDateString()` can return `null` if it’s unable to guess what the user meant (for example, `yesterday`). I don’t think we need to specify a fallback value here for that case, or for the case when `start_date` is empty, because `utilDatesOverlap()` should automatically assume the date begins in 10000 BCE. If that isn’t working, I might need to fix `utilDatesOverlap()` so it does the right thing in the validator too.

Comment (1ec5) [inline modules/renderer/features.js]: Let’s import `utilNormalizeDateString` from util.js instead of calling it from `iD`. I’m not sure if that is guaranteed to work after packaging iD.

File: modules/ui/sections/map_daterange.js
Comment (1ec5) [inline modules/ui/sections/map_daterange.js]: This can be a call to `utilNormalizeDateString()` as well: we can set `mindate_input`’s `value` to the return value’s `value`, or `DEFAULT_MIN_DATE` if the return value is `null`.

**Comment by danrademacher** [comment]
I'm looking at
http://localhost:8080/#background=Bing&daterange=0000-01-01,0000-12-31&disable_features=boundaries&id=w198794333&map=17.00/70.00269/-42.27294

And seeing this path that started in Year 30
image

Shouldn't that be filtered our by my filter to just Year 0?

File: data/core.yaml
Comment (gregallensworth) [reply data/core.yaml]: Fixed X2.

File: modules/renderer/features.js
Comment (gregallensworth) [reply modules/renderer/features.js]: I found that `utilNormalizeDateString()` can crash if `tags.start_date` or `tags.end_date` are _undefined_ or _null_ as opposed to an empty string. `utilNormalizeDateString()` goes straight into `var match = raw.match(dateRegex);` which fails if raw is undefined.

image

I expect that adding `if (! raw) return null;` to `utilNormalizeDateString()` would work around this. That would handle blank, null, and undefined.

Comment (gregallensworth) [reply modules/renderer/features.js]: Done.

Comment (1ec5) [reply modules/renderer/features.js]: Yes, feel free to make that change.

File: modules/ui/sections/map_daterange.js
Comment (1ec5) [reply modules/ui/sections/map_daterange.js]: `utilNormalizeDateString()` won’t automatically add `-mm-dd` to a bare year, but `utilDatesOverlap()` doesn’t require full precision to compare the dates correctly.

File: modules/renderer/features.js
Comment (gregallensworth) [reply modules/renderer/features.js]: I added the "if ! raw return null" to `utilNormalizeDateString()`

Then rewrote this piece of **features.js** to pick out normalized.value or else keep null, rather than specifying a fallback date.

File: modules/ui/sections/map_daterange.js
Comment (gregallensworth) [reply modules/ui/sections/map_daterange.js]: Good move. My goal had been to expand their dates to become correct. But as you say, now that we're using `utilNormalizeDateString()` and `utilDatesOverlap()` that's not strictly necessary. Changed.

**Comment by danrademacher** [comment]
I'm not reviewing this at code level but I have been testing locally, building on [my comment above](https://github.com/OpenHistoricalMap/iD/pull/142#issuecomment-1266043782).

In this view, the [Hotel Seattle](https://www.openhistoricalmap.org/way/198553375#map=20/47.60199/-122.33285&layers=O&date=1922-01-01&daterange=1922-01-01,2022-12-31) in fact disappears, as it should.

But all these other things hand around:
 image

I am digigng into it to understand why...

**Comment by 1ec5** [comment]
@danrademacher I think that’s due to https://github.com/OpenHistoricalMap/iD/pull/142#discussion_r986223288; the fix hasn’t been pushed yet.

File: modules/ui/sections/map_daterange.js
Comment (1ec5) [inline modules/ui/sections/map_daterange.js]: Flushing (resetting) the context here causes OpenHistoricalMap/issues#365 and OpenHistoricalMap/issues#447. By comparison, the checkboxes under the Map Features section adjust the features’ hidden state but don’t toss any context:

https://github.com/OpenHistoricalMap/iD/blob/e9bdd79887c5c8272c3ac9d345aa10943adaf275/modules/ui/sections/map_features.js#L119-L121 https://github.com/OpenHistoricalMap/iD/blob/e9bdd79887c5c8272c3ac9d345aa10943adaf275/modules/renderer/features.js#L305-L310 https://github.com/OpenHistoricalMap/iD/blob/e9bdd79887c5c8272c3ac9d345aa10943adaf275/modules/renderer/features.js#L54-L69

If I’m not mistaken, it’s the `redraw` event at the end of this function that causes the canvas to redraw without the newly hidden features, but nothing ever changes in terms of the object model.

Comment (gregallensworth) [reply modules/ui/sections/map_daterange.js]: Good call. I created a new `redraw()` wrapper in features.js, so `context.features().redraw();` now exists as an interface. The redrawing is MUCH FASTER now, and it sounds as if it will also correct some bugs. So that's great. Thank you.

**Comment by gregallensworth** [comment]
> I'm looking at http://localhost:8080/#background=Bing&daterange=0000-01-01,0000-12-31&disable_features=boundaries&id=w198794333&map=17.00/70.00269/-42.27294
> And seeing this path that started in Year 30 image
> Shouldn't that be filtered our by my filter to just Year 0?

Yes! I have the filtering really working this time. I was referring to the wrong commit in my notes yesterday, chased a red herring for some time.

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js]: Remember to remove this debug statement before merging.

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js:533]: I think this check is intended to prevent the filters from operating on non-OHM features, such as features in custom data layers.

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js:482]: These checks might’ve been a performance optimization, since the UI redraws pretty frequently. If `hidden()` were to include the features for which `featureFitsDateRange()` returns `false`, then `update()` will automatically add those features to `_hidden`. That would make the check here work correctly and obviate the call to `featureFitsDateRange()` just below here.

https://github.com/OpenHistoricalMap/iD/blob/f200cb225a029fd2f85f5fce3baebd39a6cf68af/modules/renderer/features.js#L251-L256

File: modules/ui/sections/map_daterange.js
Comment (1ec5) [inline modules/ui/sections/map_daterange.js:83]: `update()` has similar code for updating the URL parameters. Should we consolidate the two code paths? I think `update()` can get called in response to something other than fiddling with the checkboxes in the Map Features pane, such as panning to a place with so many points that iD filters out points automatically. It also seems to get called on startup, as soon as the map renders for the first time.

`update()` also updates a preference that gets remembered across sessions, which seems useful (but we can save that for later).

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js]: The space after the `!` results in a linter error. Can you run `npm run lint` and fix the warnings and errors? Thanks! (We’ll need to get CI set up for this fork at some point.)

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js:549]: Several tests are failing in this method:

```
Chrome Headless 108.0.5327.0 (Mac OS 10.13.6) iD.rendererFeatures hiding hides child vertices on a hidden way FAILED
	TypeError: Cannot read properties of undefined (reading '0')
	    at features2.featureFitsDateRange (dist/iD.js:45619:47 <- dist/iD.js:442:24514)
	    at features2.isHiddenFeature (dist/iD.js:45571:22 <- dist/iD.js:442:21064)
	    at features2.isHidden (dist/iD.js:45611:14 <- dist/iD.js:442:24097)
	    at features2.isHiddenChild (dist/iD.js:45587:24 <- dist/iD.js:442:22431)
	    at Context.<anonymous> (test/spec/renderer/features.js:546:29)
Chrome Headless 108.0.5327.0 (Mac OS 10.13.6): Executed 1490 of 2046 (1 FAILED) (skipped 15) (0 secs / 1.992 secs)
Chrome Headless 108.0.5327.0 (Mac OS 10.13.6) iD.rendererFeatures hiding hides child vertices on a hidden way FAILED
	TypeError: Cannot read properties of undefined (reading '0')
	    at features2.featureFitsDateRange (dist/iD.js:45619:47 <- dist/iD.js:442:24514)
	    at features2.isHiddenFeature (dist/iD.js:45571:22 <- dist/iD.js:442:21064)
	    at features2.isHidden (dist/iD.js:45611:14 <- dist/iD.js:442:24097)
	    at features2.isHiddenChild (dist/iD.js:45587:24 <- dist/iD.js:442:22431)
Chrome Headless 108.0.5327.0 (Mac OS 10.13.6) iD.rendererFeatures hiding hides uninteresting (e.g. untagged or "other") member ways on a hidden multipolygon relation FAILED
	TypeError: Cannot read properties of undefined (reading '0')
	    at features2.featureFitsDateRange (dist/iD.js:45619:47 <- dist/iD.js:442:24514)
	    at features2.isHiddenFeature (dist/iD.js:45571:22 <- dist/iD.js:442:21064)
	    at features2.isHidden (dist/iD.js:45611:14 <- dist/iD.js:442:24097)
	    at Context.<anonymous> (test/spec/renderer/features.js:574:29)
Chrome Headless 108.0.5327.0 (Mac OS 10.13.6): Executed 1491 of 2046 (2 FAILED) (skipped 15) (0 secs / 1.992 secs)
Chrome Headless 108.0.5327.0 (Mac OS 10.13.6) iD.rendererFeatures hiding hides uninteresting (e.g. untagged or "other") member ways on a hidden multipolygon relation FAILED
	TypeError: Cannot read properties of undefined (reading '0')
	    at features2.featureFitsDateRange (dist/iD.js:45619:47 <- dist/iD.js:442:24514)
	    at features2.isHiddenFeature (dist/iD.js:45571:22 <- dist/iD.js:442:21064)
	    at features2.isHidden (dist/iD.js:45611:14 <- dist/iD.js:442:24097)
Chrome Headless 108.0.5327.0 (Mac OS 10.13.6) iD.rendererFeatures hiding hides only versioned entities FAILED
	TypeError: Cannot read properties of undefined (reading '0')
	    at features2.featureFitsDateRange (dist/iD.js:45619:47 <- dist/iD.js:442:24514)
	    at features2.isHiddenFeature (dist/iD.js:45571:22 <- dist/iD.js:442:21064)
	    at features2.isHidden (dist/iD.js:45611:14 <- dist/iD.js:442:24097)
	    at Context.<anonymous> (test/spec/renderer/features.js:591:29)
Chrome Headless 108.0.5327.0 (Mac OS 10.13.6): Executed 1492 of 2046 (3 FAILED) (skipped 15) (0 secs / 1.992 secs)
Chrome Headless 108.0.5327.0 (Mac OS 10.13.6) iD.rendererFeatures hiding hides only versioned entities FAILED
	TypeError: Cannot read properties of undefined (reading '0')
	    at features2.featureFitsDateRange (dist/iD.js:45619:47 <- dist/iD.js:442:24514)
	    at features2.isHiddenFeature (dist/iD.js:45571:22 <- dist/iD.js:442:21064)
	    at features2.isHidden (dist/iD.js:45611:14 <- dist/iD.js:442:24097)
```

File: modules/renderer/features.js
Comment (gregallensworth) [reply modules/renderer/features.js]: It's commented out, but I have removed it anyway.

File: modules/renderer/features.js
Comment (gregallensworth) [reply modules/renderer/features.js]: Done.

File: modules/renderer/features.js
Comment (gregallensworth) [reply modules/renderer/features.js:533]: Replaced. See next note too.

File: modules/renderer/features.js
Comment (gregallensworth) [reply modules/renderer/features.js:482]: It is a performance optimization: if all Map Features checkboxes are checked (`_hidden` = unchecked boxes = disabled pass rules), then this feature is not hidden. But since Date Range is a different filter outside of the Map Features checkboxes, that is not accurate.

I think you're mistaken about `hidden()` It accepts a rule key (e.g. *service_roads*) and returns whether those features' rule is to be skipped. The tests for each feature are `features.isHiddenFeature` `features.isHiddenChild` `features.hasHiddenConnections` and `features.isHidden`

But how about this:
- I have reinstated the `if (!_hidden.length) return false;` conditions, which would make the rules processing faster when all boxes are checked
- then put `featureFitsDateRange()` before that test, in each of these four places

File: modules/renderer/features.js
Comment (gregallensworth) [reply modules/renderer/features.js:549]: During the tests I see that `context.features().dateRange` is undefined, thus `utilNormalizeDateString(context.features().dateRange[0]` is the "reading property of undefined"

I have added to `featureFitsDateRange` a bail condition: `if (!context.features().dateRange) return true;`

This addresses the test, as well as a hypothetical condition in which the Date Range control is removed.

File: modules/ui/sections/map_daterange.js
Comment (gregallensworth) [reply modules/ui/sections/map_daterange.js:83]: I did give a try to that, both back when I did the first version of this and also on this new version. But the results are not great: the hash will update, but not be read when the Date Range is initialized. I could continue digging, but it seems a very small return, that we would be intentionally mixing up the responsibilities -- the `daterange` URL param is set by uiSectionDateRange and it is uiSectionDateRange which reads it back on startup.

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js]: This method is called to prevent, for example, moving a node that’s connected to a hidden feature. I’d think that whether the passed-in entity falls within the date range would be irrelevant to whether it’s connected to other hidden features. Granted, this edge case would only present itself in one edge case – the one where I tried to work around OpenHistoricalMap/issues#459 by entering a way ID into the search bar.

File: modules/renderer/features.js
Comment (1ec5) [reply modules/renderer/features.js:482]: Ah, you’re right. In the future, if we allow clearing the start and end date fields without a fallback value (as with the photo overlay date filter), then `featureFitsDateRange()` could achieve a similar optimization by bailing if the dates are unset.

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js]: I wonder if we even need to normalize these dates. If a feature is tagged with a malformed date, `utilDatesOverlap()` would treat it as a blank date and assume it extends to the beginning or end of time. As a result, the feature would be shown no matter the selected date range. That would be an incentive to address the validator error.

File: modules/ui/sections/map_daterange.js
Comment (1ec5) [inline modules/ui/sections/map_daterange.js]: If the user enters a malformed but guessable date, this PR currently relies on `featureFitsDateRange()` to convert it to that guess behind the scenes. This is fine for `5 October 2022`, but for something like `5/10/2022`, it’s unclear to the user whether it interpreted the input correctly. If we unconditionally replace the `value` with `utilNormalizeDateString()`’s `value`, then it becomes more transparent to the user.

Incidentally, this would actually ensure valid inputs, so that `featureFitsDateRange()` won’t have to perform the same normalization every time it evaluates a feature.

**Review by 1ec5 — APPROVED** [review]
Thanks for iterating on this PR. I think the only thing remaining is to see if we can minimize the number of times `utilNormalizeDateString()` gets called within `featureFitsDateRange()` (which is called frequently), partly by offloading some of that to `ensureValidInputs()`.

Other than that, this should be good to merge. @danrademacher, have you gotten a chance to try it out again? I think it’s feeling pretty good at this point, but you seem to have the more extensive testbed. 😉

File: modules/renderer/features.js
Comment (gregallensworth) [reply modules/renderer/features.js]: > within the date range would be irrelevant to whether it’s connected to other hidden features

So long as we remember that "hidden" simply means "does not match a checkbox" and not "has been suppressed from the map."

I have removed this line per your instruction. This means that `hasHiddenConnections()` will instantly return false if all Map Features boxes are checked.

File: modules/ui/sections/map_daterange.js
Comment (gregallensworth) [reply modules/ui/sections/map_daterange.js]: > we unconditionally replace the value with utilNormalizeDateString()’s value, then it becomes more transparent to the user

I gotta say, the visual effect here is nice. I enter a date, and it is visibly changed and displayed in front of me. This was a good idea.

File: modules/renderer/features.js
Comment (gregallensworth) [reply modules/renderer/features.js]: > wonder if we even need to normalize these dates

Good point. I have removed these, so `featureFitsDateRange()` passes them directly to `utilDatesOverlap()` and it does seem to work as expected.

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js]: Sorry for the back-and-forth on this method. What I meant previously is that the date filter should affect whether the `connections` below here are considered hidden, not whether the passed-in `entity` is considered hidden. So if you have all the Map Features boxes checked and the feature being evaluated is inside the selected date range, but it’s connected to a feature that’s outside the selected date range, then this method should return false because that connected feature isn’t visible.

For that to happen, it needs to be possible to reach the `isHidden()` call toward the bottom of this method. So both this check and the `!_hidden.length` check would need to be removed, like you had done originally the other day. You were right to take this approach for this particular method, as opposed to the other ones.

Fortunately, this method is only called once when attempting to perform an operation on a selected feature, such as moving, but not when filtering features for rendering. So I don’t think either of these checks would result in a meaningful performance gain anyways.

**Review by 1ec5 — COMMENTED** [review]
We’re so close – thank you for your detailed work on this feature. I can’t approve the PR, so here’s some final feedback while we wait for an approval.

File: modules/renderer/features.js
Comment (gregallensworth) [reply modules/renderer/features.js]: Gotcha. Done.

**Comment by batpad** [comment]
Just wanted to show my appreciation here 👏 this is fantastic work @1ec5 @gregallensworth - can't wait to see this go live.

File: modules/ui/sections/map_daterange.js
Comment (1ec5) [inline modules/ui/sections/map_daterange.js]: The tooltip is showing up as a stringified function:

tooltip

I was mistaken in https://github.com/OpenHistoricalMap/iD/pull/142#discussion_r986157642 – `t.append()` needs to be called in situations when `t()` used to be passed directly into `html()`, but for an attribute that can’t contain HTML, `t()` is safe.

File: data/core.yaml
Comment (1ec5) [inline data/core.yaml:952]: I went to fix a typo but wound up swapping the example and format between the tooltip and placeholder. This makes the placeholder consistent with the placeholder in the Start Date and End Date fields of the inspector. I also added another example to illustrate negative years, since some mappers might be unaware of the potential for off-by-one error caused by Year 0.

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js:549]: Previously this method was implemented as a rule, which had a nice side effect that the blue chip in the status bar would include the number of features hidden for being out of range. There’s a lot of code around deciding when to recalculate the count. Patching up `gatherStats()` might do the trick.

File: modules/renderer/features.js
Comment (1ec5) [inline modules/renderer/features.js]: If you select a feature then enable a filter, the feature remains visible until you deselect the feature. This isn’t happening for date-filtered features because this check is coming after the call to `featureFitsDateRange()`.

nodes

File: modules/renderer/features.js
Comment (1ec5) [reply modules/renderer/features.js:482]: You were right about removing the `_hidden.length` checks. With them in place, vertices of hidden features would become unhidden if all the map features were enabled. I missed that case because I have Boundaries disabled by default.

**Review by 1ec5 — APPROVED** [review]
This feature seems to be working as advertised now, and it’s looking nicely polished thanks to @gregallensworth’s persistence. I’m going to keep kicking the tires on this one just in case, but we can address any remaining issues that come up in separate PRs.

Since we ended up implementing the filter as a special case instead of just another rule, there are changes scattered throughout features.js. We’ll need to be careful about testing this functionality the next time we do a major merge from upstream. I’ve added a cursory test of the date filtering functionality, but it doesn’t cover all the edge cases.

**Pull request merged**
